### PR TITLE
Fix lexically aware template boundaries and line terminators

### LINKED ISSUE #279: feat: Lexically-aware ${...} boundary detection in template literal segmentation
## Summary

The brace-counting used to find the closing `}` of a template interpolation expression (`${...}`) in `SplitTemplateAtBoundaries` (and previously in `ParseTaggedTemplate`) is **not lexically aware** — it does not account for `{`/`}` inside string literals, comments, or nested templates within the expression.

## Failure cases

- `` `${"}"}`  `` extracts `"` instead of `"}"`
- `` `${{a:1}}`  `` extracts `{a:1` instead of `{a:1}`
- `` `${/*}*/x}`  `` stops at `}` inside comment

The extracted expression is then passed to `ParseInterpolationExpression`, which fails with confusing syntax errors rather than the expected parse.

## Root cause

`ScanTemplate` in `units/Goccia.Lexer.pas` stores the **entire template body** (including `${...}` expression content) as a single flat token lexeme, with no expression-level lexical analysis. By the time the parser runs, the escape/string/comment context needed for correct brace matching is no longer available.

## Proposed fix

**Lexer-time segmentation**: Detect and record `${...}` boundaries during `ScanTemplate` while the lexer still has full lexical context (escape sequences, string delimiters, comment start/end), and embed boundary markers (e.g., `#2`) in the token lexeme to delineate static string segments from interpolation expression texts.

This is the approach described as option 1 in issue #270.

## Context

This is a known pre-existing limitation that predates PR #278. PR #278 fixed the `\${` escape boundary detection bug (where `\${` was mistakenly treated as a real interpolation boundary). The naive brace counter was intentionally carried forward in PR #278 as-is, pending this follow-up.

**References:**
- PR #278: https://github.com/frostney/GocciaScript/pull/278
- Review comment: https://github.com/frostney/GocciaScript/pull/278#discussion_r3069170312
- Requested by: @frostney

### PR DESCRIPTION
**Pull request by frostney**
## Summary
- Made template interpolation parsing lexically aware so `}` inside strings, comments, objects, and nested templates no longer terminates `${...}` early.
- Extended the lexer to recognize CR, CRLF, LS, and PS as line terminators in whitespace, hashbangs, and comments.
- Removed dynamic `import()` support from the parser/compiler path and updated the related docs and tests accordingly.
- Added lexer and template literal regression coverage for the new boundary and line-terminator behavior.

Fixes #279 

## Testing
- Not run in this turn.
- Added/updated JavaScript coverage under `tests/language/expressions/string/` for template interpolation boundary cases.
- Added lexer unit coverage in `units/Goccia.Lexer.Test.pas` for CR, CRLF, LS, and PS handling.

### CODE REVIEW

File: units/Goccia.Lexer.pas [hunk 1]
```diff
@@ -557,6 +635,304 @@ procedure TGocciaLexer.ProcessTemplateEscapeSequence(var ASB: TStringBuffer;
...
+        ASB.AppendChar(C);
+        ARawSB.AppendChar(C);
+      end;
+    else
+      begin
+        Advance;
+        ASB.AppendChar(C);
+        ARawSB.AppendChar(C);
+      end;
+    end;
```
[the remark this answers is not in the document]
Comment (frostney) [reply units/Goccia.Lexer.pas:747]: Good catch on both points. Fixed in 8922c1fd:

1. **CR column tracking**: Changed `FColumn := 0` to `FColumn := 1` in all CR branches in both `ScanNestedTemplateInExpression` and `ScanInterpolationExpression` (including the string-scanning and block-comment sub-loops). This now matches the pattern in `SkipWhitespace`, `SkipBlockComment`, and `ConsumeUnicodeLineTerminator`.

2. **LS/PS handling**: Added `UTF8_LINE_TERMINATOR_LEAD_BYTE` case branches in both methods, using the same byte-copying + `FColumn := 1` pattern as `ConsumeUnicodeLineTerminator`. Falls through to normal character handling when the lead byte isn't actually LS/PS.

Note: the existing `ScanTemplate` has the same `FColumn := 0` inconsistency for CR (line 1057) and LS/PS (line 1081) — that predates this PR and is out of scope here, but worth a follow-up.

File: units/Goccia.Lexer.pas [hunk 2]
```diff
@@ -557,6 +635,304 @@ procedure TGocciaLexer.ProcessTemplateEscapeSequence(var ASB: TStringBuffer;
...
+        ASB.AppendChar(#10);
+        ARawSB.AppendChar(#10);
+      end;
+      #10:
+      begin
+        Inc(FLine);
+        FColumn := 0;
+        Advance;
+        ASB.AppendChar(C);
+        ARawSB.AppendChar(C);
```
[the remark this answers is not in the document]
Comment (frostney) [reply units/Goccia.Lexer.pas:949]: Fixed in 8922c1fd. Changed `FColumn := 0` to `FColumn := 1` in all four CR branches within `ScanInterpolationExpression`:

1. Main expression body CR (was line 915)
2. String scanning CR (was line 809)
3. Block comment CR (was line 886)

Also added the same `FColumn := 1` pattern for LS/PS handling in the main case branch, consistent with `ConsumeUnicodeLineTerminator`.

File: units/Goccia.Parser.pas [hunk 3]
```diff
@@ -883,184 +883,83 @@ function IsValidHexString(const AValue: string): Boolean;
...
+        ACookedSegments[StaticCount] := '';
+      Inc(StaticCount);
     end
     else
     begin
-      // Normal character — advance both by 1
-      Inc(RawI);
-      Inc(CookedI);
+      AExpressionTexts[ExprCount] := Trim(RawParts[I]);
+      Inc(ExprCount);
```
[the remark this answers is not in the document]
Comment (frostney) [reply units/Goccia.Parser.pas]: Good catch — this is a real correctness issue. Fixed in 8922c1fd.

Removed `Trim()` from both `SplitTemplateAtBoundaries` and `SplitRawAtBoundaries`. Trailing newlines are semantically required to terminate `//` line comments inside interpolation expressions. Without them, `ParseInterpolationExpression` wraps the text in `(expr);` and the `;)` gets consumed by the unterminated comment.

Added a regression test for the specific failure case:
```javascript
test("line comment at end of interpolation expression", () => {
  const x = `${1 // trailing comment
}`;
  expect(x).toBe("1");
});
```

**Comment by frostney** [comment]
Addressing the second round of CodeRabbit feedback (review `4095598276`):

**Duplicate: Line comment LS/PS termination (lines 867-878)** — ✅ Already fixed in `08e69761`. Changed the loop from `(Peek <> #10) and (Peek <> #13)` to `not IsLineTerminator`, which delegates to the same `IsLineTerminator` method used by `SkipComment` and covers LF, CR, LS, and PS.

**Nitpick: Spec annotation (lines 751-763)** — ✅ Already fixed in `08e69761`. Added `// ES2026 §12.9.4 Template Literal Lexical Components` at the top of the `ScanInterpolationExpression` comment block.

**Nitpick: Shared `TEMPLATE_EXPRESSION_BOUNDARY` constant** — Intentionally kept as local constants. The sibling constants `TEMPLATE_RAW_SEPARATOR` (`#1`) and `TEMPLATE_INVALID_ESCAPE_SEPARATOR` (`#2`) follow the same local-const pattern in both files. Extracting only this one to a shared unit would be inconsistent — all three would need to move together, which is a broader refactor out of scope for this PR.
